community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.logging;
+
+import java.io.Closeable;
+
+import org.neo4j.concurrent.AsyncEvents;
+import org.neo4j.function.Consumer;
+import org.neo4j.kernel.impl.util.CappedOperation;
+import org.neo4j.kernel.impl.util.Listener;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import static org.neo4j.kernel.impl.util.CappedOperation.differentItemClasses;
+import static org.neo4j.kernel.impl.util.CappedOperation.time;
+
+/**
+ * Wraps a {@link LogProvider}, making its {@link Log logs} asynchronous by letting their log statements be
+ * treated as events fed into a {@link AsyncEvents} to be processed by a dedicated logging thread instead
+ * of in the thread making the log statement call.
+ */
+class AsyncLogProvider implements LogProvider, Closeable, Runnable
+{
+    private final AsyncEvents<AsyncLogEvent> events;
+    private final LogProvider actual;
+
+    AsyncLogProvider( LogProvider actual )
+    {
+        this( actual, new Listener<Throwable>()
+        {
+            @Override
+            public void receive( Throwable error )
+            {
+                error.printStackTrace();
+            }
+        } );
+    }
+
+    AsyncLogProvider( LogProvider actual, final Listener<Throwable> errorHandler )
+    {
+        this.events = new AsyncEvents<>( new Consumer<AsyncLogEvent>()
+        {
+            private final CappedOperation<Throwable> errorNotification = new CappedOperation<Throwable>(
+                    differentItemClasses(), time( 1, SECONDS ) )
+            {
+                @Override
+                protected void triggered( Throwable error )
+                {
+                    errorHandler.receive( error );
+                }
+            };
+
+            @Override
+            public void accept( AsyncLogEvent logEvent )
```

Comment:
Wouldn't it be a lot better if we could get a whole batch of log events here, so that we can do a single batch of writes to the underlying logging system, thus limiting the number of IO-operations we perform...